cpp/src/reductions/scan/scan_inclusive.cu

```diff
@@ -198,22 +196,210 @@ struct scan_dispatcher {
+{
+  auto const d_order_by = column_device_view::create(order_by, stream);
+  element_equality_comparator<has_nulls> element_comparator(*d_order_by, *d_order_by, true);
+  thrust::tabulate(rmm::exec_policy(stream),
+                   out.begin<size_type>(),
+                   out.end<size_type>(),
+                   [type = d_order_by->type(), element_comparator] __device__(size_type row_index) {
+                     return (row_index == 0 ||
+                             !type_dispatcher(type, element_comparator, row_index, row_index - 1))
+                              ? row_index + 1
+                              : 0;
+                   });
+}
+
+template <bool nested_nulls>
+void generate_dense_rank_struct_comparisons(column_view const& order_by,
+                                            mutable_column_view out,
+                                            rmm::cuda_stream_view stream)
+{
+  auto const d_order_by = column_device_view::create(order_by, stream);
+  thrust::tabulate(
+    rmm::exec_policy(stream),
+    out.begin<size_type>(),
+    out.end<size_type>(),
+    [d_order_by = *d_order_by, has_nulls = order_by.has_nulls()] __device__(size_type row_index) {
+      if (row_index == 0) {
+        return true;
+      } else if (has_nulls) {
+        bool const lhs_is_null{d_order_by.is_null_nocheck(row_index)};
+        bool const rhs_is_null{d_order_by.is_null_nocheck(row_index - 1)};
+        if (lhs_is_null and rhs_is_null) {
+          return false;
+        } else if (lhs_is_null != rhs_is_null) {
+          return true;
+        }
+      }
+
+      return !thrust::all_of(
+        thrust::seq,
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(0) + d_order_by.num_child_columns(),
+        [row_index, d_order_by] __device__(size_type child_index) {
+          column_device_view const& col = d_order_by.child(child_index);
+          element_equality_comparator<nested_nulls> element_comparator{col, col, true};
+          return type_dispatcher(col.type(), element_comparator, row_index, row_index - 1);
+        });
+    });
+}
+
+template <bool has_nulls>
+void generate_dense_rank_comparisons(column_view const& order_by,
+                                     mutable_column_view out,
+                                     rmm::cuda_stream_view stream)
+{
+  auto const d_order_by = column_device_view::create(order_by, stream);
+  element_equality_comparator<has_nulls> element_comparator(*d_order_by, *d_order_by, true);
+  thrust::tabulate(rmm::exec_policy(stream),
+                   out.begin<size_type>(),
+                   out.end<size_type>(),
+                   [type = d_order_by->type(), element_comparator] __device__(size_type row_index) {
+                     return (row_index == 0 ||
+                             !type_dispatcher(type, element_comparator, row_index, row_index - 1));
+                   });
+}
```

Comment:
This code is a common pattern that is often referred to as generating "head flags", i.e., finding all the points where an element is not equal to the element before it. I think it would help in understanding this function in the future to lean on the "head flags" nomenclature by renaming this function and it's result.

Reply:
Resolving because all functionality is now contained within 1 function and it does not generate + return the `head flags` as an intermediate action.